bug: nil-pointer panic in getCodeScanAlerts when REST response is nil

## Summary
`(*githubScraper).getCodeScanAlerts` dereferences `resp.StatusCode` without checking that `resp` is non-nil. When `go-github`'s `ListAlertsForRepo` returns `(nil, nil, err)` — context cancellation, DNS failure, TLS handshake failure, connection reset — the scrape goroutine panics and the entire collector pod crashes, dropping every metric recorded during that scrape (`mb.Emit` runs in the parent goroutine and never fires).

## Affected version
`ghcr.io/liatrio/liatrio-otel-collector:0.100.0` — code path still present on `main` at `receiver/githubreceiver/internal/scraper/githubscraper/helpers.go:466`.

## Steps to reproduce
1. Run the collector with `githubappauthextension` configured.
2. Enable `vcs.cve.count`.
3. Use a `search_query` broad enough to push the scrape past `timeout` (we hit this with `org:<org> archived:false` over ~300 repos, `concurrency_limit: 10`, `timeout: 14m`).

For a unit-level repro: pass a pre-cancelled `context.Context` to `getCodeScanAlerts`. `ListAlertsForRepo` returns `(nil, nil, err)` immediately and the nil-deref fires on the next line.

## Expected
The per-repo scrape logs the error and continues; other repos in the org still emit metrics.

## Actual
\`\`\`
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3d5f5dc]
goroutine 112 [running]:
...githubscraper.(*githubScraper).getCodeScanAlerts(...)
        .../helpers.go:466 +0x17c
...githubscraper.(*githubScraper).getCVEs(...)
        .../helpers.go:417 +0x6c
...githubscraper.(*githubScraper).scrape.func1()
        .../github_scraper.go:198 +0x860
\`\`\`

## Notes
Triggered in production by scrape-context timeout (surrounding log lines show \`context deadline exceeded\` from \`getDepBotAlerts\`, \`getPullRequests\`, \`getCommitData\` in the same goroutine), but any transient network condition producing a nil HTTP response will hit it.

Workaround: disable \`vcs.cve.count\`.

## Suggested fix
One-line nil guard on \`resp\` before reading \`StatusCode\`. PR to follow.

--- a/receiver/githubreceiver/internal/scraper/githubscraper/github_scraper_test.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/github_scraper_test.go
@@ -8,6 +8,7 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"path/filepath"
+	"strings"
 	"testing"
 	"time"
 
@@ -377,11 +378,12 @@ func TestScrape(t *testing.T) {
 	}
 }
 
-// panicResponses builds a mock-server response set for a single repo whose
-// code-scan alert has a nil Rule. That nil triggers a runtime panic deep
-// inside the per-repo scrape goroutine (via mapSeverities), exercising the
-// recover() that the scrape goroutine installs.
-func panicResponses(repoName string) *responses {
+// singleRepoResponses builds a happy-path mock-server response set for a
+// single repository. The panic-recovery tests below pair it with
+// panicRoundTripper to inject a synthetic panic from inside the per-repo
+// scrape goroutine, decoupling the recovery-path coverage from any specific
+// production nil-deref bug.
+func singleRepoResponses(repoName string) *responses {
 	return &responses{
 		scrape: true,
 		checkLoginResponse: loginResponse{
@@ -435,14 +437,28 @@ func panicResponses(repoName string) *responses {
 			depBotsAlerts: []VulnerabilityAlerts{{Nodes: []CVENode{}}},
 			responseCode:  http.StatusOK,
 		},
-		codeScanAlertResponse: codeScanAlertResponse{
-			codeScanAlerts: [][]*github.Alert{
-				// alert with nil Rule causes mapSeverities to nil-deref
-				{{Rule: nil}},
-			},
-			responseCode: http.StatusOK,
-		},
 	}
+}
+
+// panicRoundTripper wraps an inner http.RoundTripper and panics on any
+// request whose URL path contains trigger. It is the panic injector for the
+// recover()-path tests: the panic fires inside the per-repo scrape
+// goroutine's call to the GitHub client, which is exactly where the
+// production recover() must catch it.
+type panicRoundTripper struct {
+	inner   http.RoundTripper
+	trigger string
+}
+
+func (p *panicRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if strings.Contains(req.URL.Path, p.trigger) {
+		panic("synthetic test panic from panicRoundTripper: " + req.URL.Path)
+	}
+	inner := p.inner
+	if inner == nil {
+		inner = http.DefaultTransport
+	}
+	return inner.RoundTrip(req)
 }
 
 // TestScrapeRecoversFromPanic asserts that a runtime panic inside a per-repo
@@ -454,16 +470,15 @@ func TestScrapeRecoversFromPanic(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	server := httptest.NewServer(MockServer(panicResponses("repo1")))
+	server := httptest.NewServer(MockServer(singleRepoResponses("repo1")))
 	defer server.Close()
 
 	cfg := &Config{MetricsBuilderConfig: metadata.DefaultMetricsBuilderConfig()}
 	cfg.Metrics.VcsCveCount.Enabled = true
 
 	// Attach an observed logger so the test can confirm a panic was actually
-	// recovered. Without this guard, hardening the underlying nil-deref in
-	// mapSeverities would silently turn this test into a happy-path no-op
-	// that still passes.
+	// recovered. Without this guard, removing the panic injector below would
+	// silently turn this test into a happy-path no-op that still passes.
 	core, recorded := observer.New(zap.ErrorLevel)
 	settings := receivertest.NewNopSettings(metadata.Type)
 	settings.Logger = zap.New(core)
@@ -474,6 +489,15 @@ func TestScrapeRecoversFromPanic(t *testing.T) {
 	ghs.cfg.ConcurrencyLimit = 1000
 
 	require.NoError(t, ghs.start(ctx, componenttest.NewNopHost()))
+
+	// Inject a synthetic panic from inside the per-repo scrape goroutine by
+	// wrapping the HTTP transport. The code-scanning REST call is the last
+	// per-repo HTTP request and runs after mux.Lock(), so the panic exercises
+	// both recover() and defer mux.Unlock() in github_scraper.go.
+	ghs.client.Transport = &panicRoundTripper{
+		inner:   ghs.client.Transport,
+		trigger: "/code-scanning/alerts",
+	}
 
 	metrics, err := ghs.scrape(ctx)
 	require.NoError(t, err)
@@ -495,9 +519,10 @@ func TestScrapeLogsRecoveredPanic(t *testing.T) {
 	defer cancel()
 
 	// The mock server hard-codes "/api/v3/repos/liatrio/repo1/..." for the
-	// REST endpoints, so the failing repo must be "repo1" for the code-scan
-	// path to hit our handler.
-	server := httptest.NewServer(MockServer(panicResponses("repo1")))
+	// REST endpoints, so the failing repo must be "repo1" so all per-repo
+	// HTTP calls are served (and so panicRoundTripper has a code-scanning
+	// request to match against).
+	server := httptest.NewServer(MockServer(singleRepoResponses("repo1")))
 	defer server.Close()
 
 	cfg := &Config{MetricsBuilderConfig: metadata.DefaultMetricsBuilderConfig()}
@@ -513,6 +538,11 @@ func TestScrapeLogsRecoveredPanic(t *testing.T) {
 	ghs.cfg.ConcurrencyLimit = 1000
 
 	require.NoError(t, ghs.start(ctx, componenttest.NewNopHost()))
+
+	ghs.client.Transport = &panicRoundTripper{
+		inner:   ghs.client.Transport,
+		trigger: "/code-scanning/alerts",
+	}
 
 	_, err := ghs.scrape(ctx)
 	require.NoError(t, err)
@@ -535,11 +565,11 @@ func TestScrapeDoesNotDeadlockAfterRecoveredPanic(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	// Two repos: repo1's code-scan alert has a nil Rule and panics inside
-	// the goroutine after mux.Lock(). repo2's REST endpoints are not
-	// registered by the mock (only /repos/liatrio/repo1/... is), so its
-	// calls 404 cleanly and the goroutine reaches its own mux.Lock().
-	r := panicResponses("repo1")
+	// Two repos under ConcurrencyLimit=1 so they run sequentially. Both
+	// goroutines panic on their code-scanning HTTP call (see the
+	// panicRoundTripper installed below); the assertion is that repo2's
+	// mux.Lock() does not block on a lock leaked from repo1's panic.
+	r := singleRepoResponses("repo1")
 	r.searchRepoResponse.repos = []getRepoDataBySearchSearchSearchResultItemConnection{
 		{
 			RepositoryCount: 2,
@@ -595,6 +625,11 @@ func TestScrapeDoesNotDeadlockAfterRecoveredPanic(t *testing.T) {
 	ghs.cfg.ConcurrencyLimit = 1
 
 	require.NoError(t, ghs.start(ctx, componenttest.NewNopHost()))
+
+	ghs.client.Transport = &panicRoundTripper{
+		inner:   ghs.client.Transport,
+		trigger: "/code-scanning/alerts",
+	}
 
 	done := make(chan struct{})
 	var scrapeErr error

--- a/receiver/githubreceiver/internal/scraper/githubscraper/helpers.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/helpers.go
@@ -463,7 +463,7 @@ func (ghs *githubScraper) getCodeScanAlerts(
 	for {
 		a, resp, err := rClient.CodeScanning.ListAlertsForRepo(ctx, ghs.cfg.GitHubOrg, repo, opt)
 		if err != nil {
-			if resp.StatusCode == 404 || resp.StatusCode == 403 {
+			if resp != nil && (resp.StatusCode == 404 || resp.StatusCode == 403) {
 				ghs.logger.Sugar().Debugf("%s repo does not have any alerts or does not have alerts enabled", repo)
 				break
 			}

--- a/receiver/githubreceiver/internal/scraper/githubscraper/helpers.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/helpers.go
@@ -505,6 +505,9 @@ func mapSeverities(
 	}
 
 	for _, alert := range alerts {
+		if alert.Rule == nil || alert.Rule.SecuritySeverityLevel == nil {
+			continue
+		}
 		if val, found := mapping[strings.ToUpper(*alert.Rule.SecuritySeverityLevel)]; found {
 			m[val]++
 		}

--- a/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
@@ -1630,6 +1630,45 @@ func TestGetCVEs(t *testing.T) {
 	}
 }
 
+// Covers the 404/403 break path in getCodeScanAlerts, which previously
+// short-circuited without a nil check on resp. With the nil-guard added,
+// the response is non-nil and carries a real status, so we expect the
+// scraper to swallow the error and return nil alerts (not an error).
+func TestGetCodeScanAlertsRepoWithoutAlerts(t *testing.T) {
+	testCases := []struct {
+		desc       string
+		statusCode int
+	}{
+		{desc: "404 not found", statusCode: http.StatusNotFound},
+		{desc: "403 forbidden", statusCode: http.StatusForbidden},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(tc.statusCode)
+			}))
+			defer server.Close()
+
+			factory := Factory{}
+			defaultConfig := factory.CreateDefaultConfig()
+			settings := receivertest.NewNopSettings(metadata.Type)
+			ghs := newGitHubScraper(settings, defaultConfig.(*Config))
+			ghs.cfg.GitHubOrg = "o"
+
+			rClient := github.NewClient(nil)
+			u, err := url.Parse(server.URL + "/api-v3" + "/")
+			assert.NoError(t, err)
+			rClient.BaseURL = u
+			rClient.UploadURL = u
+
+			assert.NotPanics(t, func() {
+				alerts := ghs.getCodeScanAlerts(context.Background(), rClient, "r")
+				assert.Nil(t, alerts)
+			})
+		})
+	}
+}
+
 // Regression test for nil-pointer panic when go-github returns (nil, nil, err)
 // from ListAlertsForRepo — e.g. when the context is cancelled before the HTTP
 // response is received. Previously, getCodeScanAlerts read resp.StatusCode

--- a/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
@@ -1629,3 +1629,25 @@ func TestGetCVEs(t *testing.T) {
 		})
 	}
 }
+
+// Regression test for nil-pointer panic when go-github returns (nil, nil, err)
+// from ListAlertsForRepo — e.g. when the context is cancelled before the HTTP
+// response is received. Previously, getCodeScanAlerts read resp.StatusCode
+// without a nil check and crashed the entire collector pod.
+func TestGetCodeScanAlertsNilResponse(t *testing.T) {
+	factory := Factory{}
+	defaultConfig := factory.CreateDefaultConfig()
+	settings := receivertest.NewNopSettings(metadata.Type)
+	ghs := newGitHubScraper(settings, defaultConfig.(*Config))
+	ghs.cfg.GitHubOrg = "o"
+
+	rClient := github.NewClient(nil)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	assert.NotPanics(t, func() {
+		alerts := ghs.getCodeScanAlerts(ctx, rClient, "r")
+		assert.Nil(t, alerts)
+	})
+}

--- a/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
+++ b/receiver/githubreceiver/internal/scraper/githubscraper/helpers_test.go
@@ -1651,3 +1651,49 @@ func TestGetCodeScanAlertsNilResponse(t *testing.T) {
 		assert.Nil(t, alerts)
 	})
 }
+
+// Regression tests for nil-pointer panics in mapSeverities. GitHub's REST API
+// has been observed to omit Rule or its SecuritySeverityLevel on some alert
+// states; before the guard, the bare *alert.Rule.SecuritySeverityLevel
+// dereference crashed the entire collector pod. The mixed case also guards
+// against the nil-skip becoming an over-skip that drops well-formed alerts.
+func TestMapSeveritiesCodeScanAlerts(t *testing.T) {
+	testCases := []struct {
+		desc     string
+		alerts   []*github.Alert
+		expected map[metadata.AttributeCveSeverity]int64
+	}{
+		{
+			desc:     "nil Rule is skipped",
+			alerts:   []*github.Alert{{Rule: nil}},
+			expected: map[metadata.AttributeCveSeverity]int64{},
+		},
+		{
+			desc:     "nil SecuritySeverityLevel is skipped",
+			alerts:   []*github.Alert{{Rule: &github.Rule{SecuritySeverityLevel: nil}}},
+			expected: map[metadata.AttributeCveSeverity]int64{},
+		},
+		{
+			desc: "valid alerts are counted alongside malformed ones",
+			alerts: []*github.Alert{
+				{Rule: nil},
+				{Rule: &github.Rule{SecuritySeverityLevel: nil}},
+				{Rule: &github.Rule{SecuritySeverityLevel: github.Ptr("HIGH")}},
+				{Rule: &github.Rule{SecuritySeverityLevel: github.Ptr("CRITICAL")}},
+			},
+			expected: map[metadata.AttributeCveSeverity]int64{
+				metadata.AttributeCveSeverityHigh:     1,
+				metadata.AttributeCveSeverityCritical: 1,
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			var got map[metadata.AttributeCveSeverity]int64
+			assert.NotPanics(t, func() {
+				got = mapSeverities(nil, tc.alerts)
+			})
+			assert.Equal(t, tc.expected, got)
+		})
+	}
+}

